Add warnings acknowledgements

arcalys opened the pull request:
### Added
- Warning acknowledgments:
  - Warning level field (`requiresacknowlegement`) to control acknowledgment requirements for individual warnings.
  - New `acknowledge_warning()` method in `WarningsHandler`.
  - New `do_acknowledge` action in `usercp.php`.
  - New `warninglog` action in `usercp.php`.
  - User level field (`unacknowledgedwarnings`) for tracking the count of unacknowledged warnings per user.
- New warning log view in the user control panel (displayed if warnings points > 0).
- New warning notice view to be used in header messages.

### Changed
- Moved `expires` language key to `global.php` to avoid duplicating it.
- Updated warning-related views (e.g., ModCP) to include acknowledgment information.

### Fixed

- Incorrect display in `usercp` home view for:
  - Warning expiration (wrong field used).
  - Warning level header (HTML not parsed).

### Additional

- Testing/review welcome.
- @justinsoltesz if you see how this can be improved visually, any hint/help is appreciated

### Screenshots
Moderator view:
<img width="3840" height="2103" alt="image" src="https://github.com/user-attachments/assets/0b8590b8-7a64-4325-99e3-7267c859300e" />
<img width="3840" height="2103" alt="image" src="https://github.com/user-attachments/assets/8bc3a667-5ac7-4472-9527-9737be1f9f5d" />
<img width="3840" height="2103" alt="image" src="https://github.com/user-attachments/assets/29840528-ba3f-4ea4-a9af-032976cd901c" />
<img width="3840" height="2103" alt="image" src="https://github.com/user-attachments/assets/d44b7863-a559-4794-abcc-9874dc785ebf" />
<img width="3840" height="2103" alt="image" src="https://github.com/user-attachments/assets/4eb48513-fc11-4640-b45c-83c9fc8aa2cb" />

User view:
<img width="3840" height="2103" alt="image" src="https://github.com/user-attachments/assets/b5f2f3e0-fed1-47e2-b6db-0792ca132b39" />
<img width="3840" height="2103" alt="image" src="https://github.com/user-attachments/assets/e6e6821f-3111-4726-99c0-b1c2d283ce05" />
<img width="3840" height="2103" alt="image" src="https://github.com/user-attachments/assets/be3114c4-c836-4e08-846b-13475168eaa3" />
<img width="3840" height="2103" alt="image" src="https://github.com/user-attachments/assets/7bec9a07-7617-4c1a-b479-a1070f4bb7ad" />

Resolves #5192

arcalys committed:
Add warnings acknowledgement

dvz commented:
Perhaps instead of a global setting, a warning type default?

The boolean & time columns in `mybb_warnings` could be merged into a single `acknowledged` with time.

A cache count column in `mybb_users` should work better than datacache to avoid storing user-specific details globally.

We can expand the remaining `ack` identifiers to full word for consistency.

---

> Do we keep the `acknowledge` button/action where a warning is expired?

IMO yes

> Do we keep the acknowledge button/action where a warning is revoked?

IMO no

> Do we want to add extra action in addition to this? (e.g. prevent users from posting until all their warnings are acknowledged)?

It should be easy in core & plugins with the mentioned user table cache: `if (!user.unacknowledged_warnings)`

dvz commented on warnings.php:
```diff
@@ -484,6 +489,36 @@
 }
 
+// Acknowledge a warning
+if($mybb->input['action'] === "do_ack" )
```
This will need to be POST (#4329), and use `verify_post_check()`

arcalys commented:
> Perhaps instead of a global setting, a warning type default?
> 
> The boolean & time columns in `mybb_warnings` could be merged into a single `acknowledged` with time.
> 
> A cache count column in `mybb_users` should work better than datacache to avoid storing user-specific details globally.
> 
> We can expand the remaining `ack` identifiers to full word for consistency.
> 
> > Do we keep the `acknowledge` button/action where a warning is expired?
> 
> IMO yes
> 
> > Do we keep the acknowledge button/action where a warning is revoked?
> 
> IMO no
> 
> > Do we want to add extra action in addition to this? (e.g. prevent users from posting until all their warnings are acknowledged)?
> 
> It should be easy in core & plugins with the mentioned user table cache: `if (!user.unacknowledged_warnings)`

- Merged the database fields.
- Removed the ACK button for revoked warnings.
- Switched to POST + verify

I'll check the cache next.

Sama34 commented:
Hi,

Please consider the following:
- Remove the global `enablewarningsacknowledgements` setting, seems redundant as no further features are added for it to make a substantial difference
- Add to the `Group Promotions` system, tho can also be added in #5189 later if you think so
- Remove acronyms or shortenings like `ack_` or `acked`, it may be confusing (I myself was unsure what `ACK` meant tho it now seems obvious) It may also cause confusion with the ACK protocol

> Do we want to add extra action in addition to this? (e.g. prevent users from posting until all their warnings are acknowledged)?

Once integrated into the promotion system, this would be achieved indirectly.

Style-wise, I think it looks fine, but consider kindly the suggestions already provided by others.

Regards.

Sama34 commented on inc/datahandlers/warnings.php:
```diff
@@ -737,5 +738,22 @@ function update_warning()
+	 * Acknowledges a warning
+	 *
+	 */
+	function acknowledge() : void
```
Make sure to add a hook he, though the best location would be better out figured after you update the logic.

arcalys replied:
Added a hook.

Sama34 commented on usercp.php:
```diff
@@ -3348,6 +3351,63 @@
 
+
+if($mybb->input['action'] == "warninglog")
+{
```
Since this page will not be shown to users based on their warning level (=0), perhaps block it as well based on the same logic.

arcalys commented:
> Hi,
> 
> Please consider the following:
> 
>     * Remove the global `enablewarningsacknowledgements` setting, seems redundant as no further features are added for it to make a substantial difference
> 
>     * Add to the `Group Promotions` system, tho can also be added in [Add new Group Promotion rules #5189](https://github.com/mybb/mybb/issues/5189) later if you think so
> 
>     * Remove acronyms or shortenings like `ack_` or `acked`, it may be confusing (I myself was unsure what `ACK` meant tho it now seems obvious) It may also cause confusion with the ACK protocol
> 
> 
> > Do we want to add extra action in addition to this? (e.g. prevent users from posting until all their warnings are acknowledged)?
> 
> Once integrated into the promotion system, this would be achieved indirectly.
> 
> Style-wise, I think it looks fine, but consider kindly the suggestions already provided by others.
> 
> Regards.

I have:
- removed the global setting.
- renamed all `ack` fields and keys to `acknowledgement`.
- hidden the warnings logs menu item from user CP when warningpoints is 0
- added the `unacknowledgedwarnings` field to `mybb_user` to avoid using the global cache and update the warnings handler to update the count for `insert_warning`, `update_warning` and `acknowledge_warning`.
- added a hook to `acknowledge_warning()` as requested.

justinsoltesz commented on inc/themes/core.base/frontend/templates/modcp/warninglogs/item.twig:
```diff
@@ -16,6 +16,17 @@
+            <div class="warning-log__acknowledgement">
+                {{ lang.acknowledged }}:
+                {% if warning.acknowledged %}
+                <span class="warning-log__acknowledgement--good">{{ lang.yes }}</span>
```
This line can be deleted. The date alone is enough to indicate the warning has been acknowledged, no need to also say "Yes"

justinsoltesz commented on inc/themes/core.base/frontend/templates/modcp/warninglogs/item.twig:
```diff
@@ -16,6 +16,17 @@
+                {{ lang.acknowledged }}:
+                {% if warning.acknowledged %}
+                <span class="warning-log__acknowledgement--good">{{ lang.yes }}</span>
+                <span class="warning-log__acknowledgement--date">({{ warning.acknowledged|my_date }})</span>
```
With removal of line 23, you can delete the parentheses here

Also, CSS class is incorrect per [BEM syntax](https://getbem.com/naming/). Probably should be `warning-log__acknowledgement-date`. The double hyphen is reserved for modifiers.

justinsoltesz commented on inc/themes/core.base/frontend/templates/usercp/home.twig:
```diff
@@ -191,10 +191,34 @@
+                                {{ lang.acknowledged }}:
+                                {% if warning.acknowledged %}
+                                <span class="warning-log__acknowledgement--good">{{ lang.yes }}</span>
+                                <span class="warning-log__acknowledgement--date">({{ warning.acknowledged|my_date }})</span>
```
Same comments from `modcp/warninglogs/item.twig` apply here

justinsoltesz commented on inc/themes/core.base/frontend/templates/modcp/warninglogs/item.twig:
```diff
@@ -16,6 +16,17 @@
+                <span class="warning-log__acknowledgement--good">{{ lang.yes }}</span>
+                <span class="warning-log__acknowledgement--date">({{ warning.acknowledged|my_date }})</span>
+                {% else %}
+                <span class="warning-log__acknowledgement--bad">{{ lang.no }}</span>
```
I'm thinking something like this to bring attention to unacknowledged warnings:

<img width="956" height="143" alt="image" src="https://github.com/user-attachments/assets/13111e6e-3d24-4391-a132-47a62aeed172" />

Can be done with something like:
```
<span class="warning-log__acknowledgement-required tag tag--bad">Not Acknowledged</span>
```

Need to move the "Acknowledged:" bit within the `if warning.acknowleged` conditional.

justinsoltesz commented on inc/themes/core.base/frontend/templates/usercp/home.twig:
```diff
@@ -191,10 +191,34 @@
+                                <span class="warning-log__acknowledgement--good">{{ lang.yes }}</span>
+                                <span class="warning-log__acknowledgement--date">({{ warning.acknowledged|my_date }})</span>
+                                {% else %}
+                                <span class="warning-log__acknowledgement--bad">{{ lang.no }}</span>
```
Same comment from `modcp/warninglogs/item.twig`, except I think for users it should say "Requires Acknowledgement"

justinsoltesz commented on inc/themes/core.base/frontend/templates/usercp/warninglog.twig:
Same comments from `usercp/home.twig` apply to this file

justinsoltesz commented on inc/themes/core.base/frontend/templates/warnings/view.twig:
```diff
@@ -50,6 +63,14 @@
+                    <span class="list__field">{{ lang.acknowledgment_required }}</span>
+                    <span class="list__value">{{ warning.requiresacknowledgement ? lang.yes : lang.no }}</span>
+                </li>
+                <li class="list__item">
```
If acknowledgement isn't required, there's no need to show the acknowledgment date item.

justinsoltesz commented on inc/themes/core.base/frontend/templates/warnings/warning_row.twig:
Same comments from other template files apply

justinsoltesz reviewed (commented):
I have left suggestions that should help clean up the UI a bit

dvz commented on inc/datahandlers/warnings.php:
```diff
@@ -374,7 +374,7 @@ function update_user($method='insert')
 		}
 
-		$warning = &$this->data;
+		$warning = &$this->read_warning_data;
```
was it causing issues?

arcalys replied:
No, just a mistake

dvz commented on inc/datahandlers/warnings.php:
```diff
@@ -662,6 +662,11 @@ function update_user($method='insert')
 			}
 		}
 
+		$query = $db->simple_select("warnings", "COUNT(wid) AS count", "uid={$user['uid']} AND requiresacknowledgement=1 AND acknowledged=0");
```
other conditions factor in revocation - not included here

dvz commented on inc/languages/english/warnings.lang.php:
```diff
@@ -36,6 +36,10 @@
+$l['require_acknowledgement'] = "Require Acknowledgement";
+$l['require_acknowledgement_description'] = "Require the user to acknowledge this warning.";
+$l['acknowledgment_required'] = "Acknowledgment Required?";
+$l['acknowledgment_date'] = "Acknowledgment Date";
```
we can standardize `acknowledgment` to `acknowledgement` (more prevalent here)

dvz commented on inc/schemas/mysql_db_tables.php:
```diff
@@ -1188,6 +1189,8 @@ classicpostbit tinyint(1) NOT NULL default '0',
 	points smallint unsigned NOT NULL default '0',
 	dateline int unsigned NOT NULL default '0',
 	issuedby int unsigned NOT NULL default '0',
+	requiresacknowledgement tinyint(1) NOT NULL default '1'
```
missing `,`

dvz replied:
```suggestion
	requiresacknowledgement tinyint(1) NOT NULL default '1',
```

dvz commented on global.php:
```diff
@@ -646,6 +646,15 @@
+{
+	$headerMessages['warningsnotice']['id'] = 'warning_notice';
+	$headerMessages['warningsnotice']['class'] = 'alert--danger';
+	$headerMessages['warningsnotice']['message'] = $lang->sprintf($lang->unacked_warnings_notice, $warnings_count);
```
remaining short term `unacked` here and in the comment

dvz commented on usercp.php:
```diff
@@ -3348,6 +3351,68 @@
+	$query = "
+		SELECT COUNT(w.wid) AS count
+		FROM " . TABLE_PREFIX . "warnings w
+		LEFT JOIN " . TABLE_PREFIX . "warningtypes t ON w.tid = t.tid
```
looks like we don't need this `JOIN`

dvz commented on inc/themes/core.base/frontend/templates/layouts/account.twig:
```diff
@@ -56,6 +56,14 @@
                 <span class="section-menu__text">{{ lang.nav_subscriptions }}</span>
             </a>
         </li>
+        {% if mybb.user.warningpoints %}
```
looks like it also hides the link for inactive warnings - we could display it if any history is present (perhaps by moving the simple `COUNT` query up to be available for a check here)

dvz commented on inc/themes/core.base/frontend/templates/messages/warningsnotice.twig:
`{% extends 'messages/base.twig' %}` like the others?

arcalys replied:
Updated

dvz commented on inc/themes/core.base/frontend/templates/usercp/home.twig:
```diff
@@ -191,10 +191,34 @@
+                        {% endif %}
                     </div>
                     <div class="row__secondary">
+                        {% if warning.requiresacknowledgement and not warning.acknowledged %}
```
condition not consistent with `home.twig` (`and not warning.daterevoked`)

dvz commented on inc/datahandlers/warnings.php:
```diff
@@ -662,6 +662,11 @@ function update_user($method='insert')
 			}
 		}
 
```
add section comment for better scannability?

dvz commented on warnings.php:
```diff
@@ -484,6 +485,32 @@
 }
 
+// Acknowledge a warning
+if($mybb->input['action'] === "do_acknowledgement" && $mybb->request_method === "post")
```
```suggestion
if($mybb->input['action'] === "do_acknowledge" && $mybb->request_method === "post")
```

dvz commented on usercp.php:
```diff
@@ -3348,6 +3351,68 @@
+		$row['revoked_date'] = ($row['daterevoked'] > 0)? my_date('relative', $row['daterevoked']) : $row['daterevoked'];
+		$row['title'] = !empty($row['title']) ? $row['title'] : $row['custom_title'];
+		$row['points'] = ($row['points'] > 0) ? '+' . $row['points'] : $row['points'];
+		$row['expire_date'] = ($row['expired'] === 1) ? $lang->already_expired : (($row['expires'] > 0) ? nice_time($row['expires'] - TIME_NOW) : $lang->never);
```
some engines may return strings

```suggestion
		$row['expire_date'] = $row['expired'] ? $lang->already_expired : (($row['expires'] > 0) ? nice_time($row['expires'] - TIME_NOW) : $lang->never);
```

dvz commented on usercp.php:
```diff
@@ -3348,6 +3351,68 @@
+		$row['title'] = !empty($row['title']) ? $row['title'] : $row['custom_title'];
+		$row['points'] = ($row['points'] > 0) ? '+' . $row['points'] : $row['points'];
+		$row['expire_date'] = ($row['expired'] === 1) ? $lang->already_expired : (($row['expires'] > 0) ? nice_time($row['expires'] - TIME_NOW) : $lang->never);
+		$warning_list[] = $row;
```
looks like we can build the array directly:
```php
$warning_list[] = [
    // ...
];
```

and make ternary expressions more readable:
```php
$example = $condition
    ? true
    : false;
```

Sama34 replied:
If the array is build directly, please pass on the identifiable field `wid`.

dvz commented on inc/themes/core.base/frontend/templates/warnings/view.twig:
```diff
@@ -50,6 +63,14 @@
+                </li>
+                <li class="list__item">
+                    <span class="list__field">{{ lang.acknowledgment_date }}</span>
+                    <span class="list__value">{{ warning.acknowledged ? warning.acknowledged|my_date : lang.na  }}</span>
```
```suggestion
                    <span class="list__value">{{ warning.acknowledged ? warning.acknowledged|my_date : lang.na }}</span>
```

dvz commented on warnings.php:
```diff
@@ -484,6 +485,32 @@
+		error($lang->warning_already_acknowledged);
+	}
+
+	if(!(int)$warning['requiresacknowledgement'])
```
more readable without the casts (it's converted to `bool` for comparison anyway)?

Sama34 replied:
Maybe `empty($warning['requiresacknowledgement'])` instead.

dvz reviewed (commented):
- Twig templates can be brushed up by indenting when nested - in both Twig and HTML structures (not consistent everywhere)
- We'll also need a relevant upgrade script section

dvz commented:
We'll now be using `warnings.php` for the user action, not just moderators - there seems to be an overlap with `warnings.php?uid=?` (Mod CP also has a _warning log_, but it's more broad).

Perhaps we can simplify the logic:
- use `warnings.php` defaulting to own account for listing
- use `warnings.php?action=view&wid=?` to inspect, and access the action link
- remove the action button from other places
- change the header message link to the `warnings.php` listing

Then, internally and UX-wise, there would be no _warning log_ for users - just _Warnings_.

It would need a review of access & data permissions.

Sama34 commented on inc/datahandlers/warnings.php:
```diff
@@ -737,5 +743,22 @@ function update_warning()
+		global $db, $plugins;
+
+		$warning = &$this->read_warning_data;
+		$ack_time = TIME_NOW;
```
This variable seems redundant.

arcalys commented:
> * Twig templates can be brushed up by indenting when nested - in both Twig and HTML structures (not consistent everywhere)
> 
>     * We'll also need a relevant upgrade script section

- Updated the indentation.
- Added the upgrade script.

arcalys commented:
> I have left suggestions that should help clean up the UI a bit

I think I’ve addressed everything you asked for, but feel free to point out anything I may have missed.

justinsoltesz commented on inc/themes/core.base/frontend/styles/components/_tags.scss:
```diff
@@ -9,6 +9,7 @@
 
     &--float {
         float: $right;
+        margin-left: 3px;
```
this needs to be `margin-#{$left}: 3px;`

arcalys replied:
Updated

Sama34 commented on usercp.php:
```diff
@@ -3448,6 +3528,36 @@
+{
+	verify_post_check($mybb->get_input('my_post_key'));
+
+	require_once MYBB_ROOT.'inc/datahandlers/warnings.php';
```
```suggestion
	require_once MYBB_ROOT.'inc/functions_warnings.php';
	require_once MYBB_ROOT.'inc/datahandlers/warnings.php';
```

Sama34 committed:
Update usercp.php

Sama34 committed:
Merge branch 'dev-1.9' into feat/5192-warnings-ack
